### ISSUE TITLE
Invalid assertion on two dimension array value with non-determinstic index

### ISSUE THREAD

**Issue by guhetier** [esbmc/esbmc#3]
ESBMC detect an invalid assertion violation when : 

- the assertion condition is about a value in a two dimension array
- an index get a non-deterministic value, that is forced to a specific value by pruning all values but one

ESBMC seems to process correctly assertions about the value of the non-deterministic index alone, or
about the value in the array with constant indexes.

```c
/* esbmc esbmc_array_nondet_err.c */
int main(void)
{
    unsigned int b;
    unsigned int a = 0;
    int table[2][2]  = {{1,0}, {1,0}};

    if (nondet_uint() == 0) {
        a = 1;
    } else {
        __ESBMC_assume(0);
    }
    __ESBMC_assert(table[1][0], "FOO"); /* Assertion ok */
    __ESBMC_assert(a == 1, "FOO"); /* Assertion ok */
    b = table[a][0];
    __ESBMC_assert(b, "ERROR"); /* Assertion violated */
    __ESBMC_assert(table[a][0], "ERROR"); /* arrays bound violated */
    return (0);
}
```

Trace : 

```
/*
ESBMC version 4.2.0 64-bit x86_64 linux
file esbmc_array_nondet_err.c: Parsing
Converting
Type-checking esbmc_array_nondet_err
Generating GOTO Program
GOTO program creation time: 0.074s
GOTO program processing time: 0.000s
Starting Bounded Model Checking
Symex completed in: 0.001s (27 assignments)
Slicing time: 0.000s
Generated 6 VCC(s), 5 remaining after simplification
No solver specified; defaulting to Boolector
Encoding remaining VCC(s) using bit-vector arithmetic
Encoding to solver time: 0.003s
Solving with solver Boolector 2.4.1
Encoding to solver time: 0.003s
Runtime decision procedure: 0.000s
Building error trace

Counterexample:

State 1 file esbmc_array_nondet_err.c line 8 function main thread 0
c::main
----------------------------------------------------
  c::main::$tmp::return_value_nondet_uint$1=0 (0)

State 4 file esbmc_array_nondet_err.c line 17 function main thread 0
c::main
----------------------------------------------------
  esbmc_array_nondet_err::main::1::b=0 (0)

State 5 file esbmc_array_nondet_err.c line 18 function main thread 0
c::main
----------------------------------------------------
Violated property:
  file esbmc_array_nondet_err.c line 18 function main
  ERROR
  (_Bool)b

VERIFICATION FAILED
*/
```

**Comment by mikhailramalho** [comment]
The bug is related to our array flattener code and how we convert the indexes. 

The fix will be released as part of ESBMC 4.2.2.

Thank you for the report.

**Comment by mikhailramalho** [comment]
Hi,

The fixes were released as part of ESBMC 4.3.
